src/time_manager.h

```diff
@@ -0,0 +1,169 @@
+#pragma once
+
+#include "bit_board.h"
+#include "board_defs.h"
+#include <atomic>
+#include <cstdint>
+#include <optional>
+
+class TimeManager {
+public:
+    static inline bool timeForAnotherSearch(const BitBoard& board)
+    {
+        if (s_timedOut)
+            return false;
+
+        /* always allow full scan on first move - will be good for the hash table :) */
+        if (board.fullMoves > 0) {
+            const auto now = std::chrono::system_clock::now();
+            const auto timeLeft = s_endTime - now;
+            const auto timeSpent = now - s_startTime;
+
+            /* factor is "little less than half" meaning that we give juuust about half the time we spent to search a new depth
+             * might need tweaking - will do when game phases are implemented */
+            const auto timeLimit = timeSpent / 1.9;
+
+            if (timeLeft < timeLimit) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    static inline void start(const BitBoard& board)
+    {
+        s_startTime = std::chrono::system_clock::now();
+        setupTimeControls(board);
+        s_timedOut = false;
+    }
+
+    static inline void startInfinite()
+    {
+        s_startTime = std::chrono::system_clock::now();
+        s_endTime = TimePoint::max();
+        s_timedOut = false;
+    }
+
+    /* should only be called from searchers
+     * will check for timeout from the searchers themselves
+     * this will remove the need for a time thread or similar
+     * and will make the engine more deterministic */
+    static inline bool searcherCheckForTimeout(uint64_t nodes)
```

Comment:
If this is only to be called from the searchers, does it not belong in the searcher class?

Reply:
I agree :+1: will move the check to the searcher instead :)
Was trying to have minimal impact on the searcher when I implemented this initially (to avoid rebase chaos) :)